Make Apple Validator hit production then sandbox

Timothylock opened the pull request:
Call Production App Store First. Then call sandbox. We're currently
doing this outside of the library which isn't ideal.
https://developer.apple.com/library/content/technotes/tn2259/_index.html

How do I verify my receipt (iOS)?

Always verify your receipt first with the production URL; proceed to
verify with the sandbox URL if you receive a 21007 status code.
Following this approach ensures that you do not have to switch between
URLs while your application is being tested or reviewed in the sandbox
or is live in the App Store.

PLAT-6523

Commit:
Add TryBothStores config option

Call Production App Store First. Then call sandbox. We're currently
doing this outside of the library which isn't ideal.
https://developer.apple.com/library/content/technotes/tn2259/_index.html

How do I verify my receipt (iOS)?

Always verify your receipt first with the production URL; proceed to
verify with the sandbox URL if you receive a 21007 status code.
Following this approach ensures that you do not have to switch between
URLs while your application is being tested or reviewed in the sandbox
or is live in the App Store.

PLAT-6523

Commit:
Correct ignoring err

Error should not be ignored

PLAT-6523

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Wattpad/go-iap?pullRequest=1) <br/>All committers have signed the CLA.

Commit:
Optimize lines of code

TryBothStores: false, not needed since bool by default is false.
Don’t even check for err unless TryBothStores is true. Short circuit
faster in some cases.

Commit:
Logic fix

Read out a separate body complete separate from the result interface
and just unmarshal to result if unmarshalling to IAPResponse fails.

PLAT-6523

Commit:
Reduce nested if else statement

Left out a case where a prod receipt hits it.

PLAT-6523

Timothylock commented:
Adding this feature as an optional feature because we don't want to break the original functionality of the library. There are times where one might only want to hit against one store.

richpoirier reviewed (changes requested):
I would actually make checking production, then the sandbox the default behaviour. That's what's recommended by Apple.

That means you wouldn't need an `IsProduction` config var anymore either which simplifies it (why I think it's the right thing to do).

`Client` and `URL` are still exported so anyone can still set the url to whatever they want for testing.

Hopefully the original author agrees and accepts our PR but if not we could just keep using our fork.

Commit:
Make Prod -> Sandbox the default behaviour

Make Prod -> Sandbox the default behaviour as that is what is
recommended by Apple.

PLAT-6523

richpoirier commented on appstore/validator.go:
```diff
@@ -120,7 +121,20 @@ func (c *Client) Verify(req IAPRequest, result interface{}) error {
 	}
 	defer resp.Body.Close()
 
-	err = json.NewDecoder(resp.Body).Decode(result)
```
You could keep this and do a type assertion instead:

```go
r, ok := result.(IAPResponse)
```

Timothylock replied:
This line is moved to line 148. The line still exists. I believe the reason that `Verify` takes in an interface{} is to be flexible. Thats why the only time we need an IAPResponse is to check for 21007, but it should not return an error if there is an error (meaning that the error should not affect the outcome of the function)

richpoirier replied:
You're now unmarshaling twice even if you don't get a status code 21007.

`json.NewDecoder` is great for when you have a reader, it'll stream it into a buffer for you (instead of you using `ioutil.ReadAll()` to do it yourself).

It's hard to explain in words but this is what I would do:

```golang
	err = json.NewDecoder(resp.Body).Decode(result)
	if err != nil {
		return err
	}

	// Always verify your receipt first with the production URL; proceed to verify with the sandbox URL if you receive
	// a 21007 status code
	//
	// https://developer.apple.com/library/content/technotes/tn2413/_index.html#//apple_ref/doc/uid/DTS40016228-CH1-RECEIPTURL
	r, ok := result.(IAPResponse)
	if ok && r.Status == 21007 {
		resp, err := client.Post(SandboxURL, "application/json; charset=utf-8", b)
		if err != nil {
			return err
		}
		defer resp.Body.Close()

		return json.NewDecoder(resp.Body).Decode(result)
	}

	return nil
```

This won't change the result of the function. If it couldn't assert the type as `IAPResponse`, like before, it won't return an error.

richpoirier commented on appstore/validator.go:
```diff
@@ -120,7 +121,20 @@ func (c *Client) Verify(req IAPRequest, result interface{}) error {
+
+	r := &IAPResponse{}
+	err = json.Unmarshal(bodyBytes, r)
+	if err == nil && r.Status == 21007 {
```
Definitely want a comment here explaining why we're doing this, ideally with a link to Apple's docs.

richpoirier commented on appstore/validator.go:
```diff
@@ -120,7 +121,20 @@ func (c *Client) Verify(req IAPRequest, result interface{}) error {
+	err = json.Unmarshal(bodyBytes, r)
+	if err == nil && r.Status == 21007 {
+		c.URL = SandboxURL
+		err = c.Verify(req, result)
```
Try to avoid recursion unless the alternative is much more difficult or less efficient. In this case, without recursion you'd be able to reuse the http client and the request body. Then just do another POST right here with the sandbox URL. Then you don't have to set the URL back after, which seems a little hacky.

richpoirier commented:
Probably want a new title for this PR now

Commit:
Add comment + No recursion

Add comment to explain changes
No recursion since client can be reused

PLAT-6523

Commit:
Enforce type

Enforce the type and attempt to locate a 21007

PLAT-6523

Timothylock commented:
I see what you mean now and I noticed that error. Corrected based on your suggestion!

Not too sure if the second defer body close is needed since there was already a line in the top that defers the closing of the body.

richpoirier reviewed (approved):
I actually don't know the answer to whether the second defer Close() is needed.

Exercise for the reader: make a small program that demonstrates the difference between one defer close, and two 😉 (would definitely give you a better understanding of how the lower levels of Go).
